autoload/ale/code_lens.vim

```diff
@@ -0,0 +1,185 @@
+    let l:namespace = s:GetNamespace(a:buffer)
+    for item in a:items
+        let text = '  ' . substitute(l:item.title, '\n', ' ', 'g')
+        call nvim_buf_set_virtual_text(
+        \ a:buffer, l:namespace,
+        \ item.range.start.line, [[text, 'Comment']], {}
+        \)
+    endfor
+endfunction
+
+function! ale#code_lens#Clear(buffer) abort
+    let l:namespace = s:GetNamespace(a:buffer)
+    call nvim_buf_clear_namespace(a:buffer, l:namespace, 0, -1)
+endfunction
+
+function! ale#code_lens#HandleLSPResponse(buffer, conn_id, response) abort
+    if has_key(a:response, 'id')
+    \&& has_key(s:code_lens_map, a:response.id)
+        let l:options = remove(s:code_lens_map, a:response.id)
+
+        let l:result = get(a:response, 'result', v:null)
+        let l:item_list = []
+
+        if type(l:result) is v:t_list
+            " Each item looks like this:
+            "{
+            "  'range': {
+            "    'start': { 'line': 7, 'character': 2 },
+            "    'end': { 'line': 7, 'character': 12 }
+            "  },
+            "  'command': { 'title': 'int', 'command': '' }
+            "}
+            for l:response_item in l:result
+                call add(l:item_list, {
+                \ 'range': l:response_item.range,
+                \ 'title': l:response_item.command.title,
+                \})
+            endfor
+        endif
+
+        if !empty(l:item_list)
+            call ale#code_lens#Show(a:buffer, l:item_list)
+        else
+        endif
+    endif
+endfunction
+
+function! s:OnReady(linter, lsp_details, ...) abort
+    let l:buffer = a:lsp_details.buffer
+
+    " If we already made a request, stop here.
+    if getbufvar(l:buffer, 'ale_code_lens_request_made', 0)
+        return
+    endif
+
+    let l:id = a:lsp_details.connection_id
+
+    let l:Callback = function('ale#code_lens#HandleLSPResponse', [l:buffer])
+    call ale#lsp#RegisterCallback(l:id, l:Callback)
+
+    let l:message = ale#lsp#message#CodeLens(l:buffer)
+    let l:request_id = ale#lsp#Send(l:id, l:message)
+
+    call setbufvar(l:buffer, 'ale_code_lens_request_made', 1)
+    let s:code_lens_map[l:request_id] = {
+    \   'buffer': l:buffer,
+    \}
+endfunction
+
+function! s:Request(linter, buffer) abort
+    let l:lsp_details = ale#lsp_linter#StartLSP(a:buffer, a:linter)
+
+    if !empty(l:lsp_details)
+        call ale#lsp#WaitForCapability(
+        \   l:lsp_details.connection_id,
+        \   'code_lens',
+        \   function('s:OnReady', [a:linter, l:lsp_details]),
+        \)
+    endif
+endfunction
+
+function! ale#code_lens#Request(buffer) abort
+    if !ale#code_lens#IsEnabled(a:buffer)
+        return
+    endif
+
+    " Set a flag so we only make one request.
+    call setbufvar(a:buffer, 'ale_code_lens_request_made', 0)
+
+    for l:linter in ale#linter#Get(getbufvar(a:buffer, '&filetype'))
+        if !empty(l:linter.lsp) && l:linter.lsp isnot# 'tsserver'
+            call s:Request(l:linter, a:buffer)
```

Comment:
This is incorrect now as requests to different LSPs will override each others codelens, I'm not sure though if we need to support the case with multiple LSPs + codelens, thoughts?

Reply:
You can either combine the results of multiple servers, or ignore responses from other servers. Note that if you choose to ignore responses, you should make a request to all available servers, but keep responses from only one.